Add person, student and teacher classes

leonard33 opened the pull request:
Create classes and inheritance:
Create class Person with the following:
Instance vars: @id, @name, and @age.
Constructor with name, age, and parent_permission as parameter. name and parent_permission are optional and have default values of "Unknown" and true.
Getters for @id, @name, and @age.
Setters for @name and @age.
Private method is_of_age? that returns true if @age is greater or equal to 18 and false otherwise.
Public method can_use_services? that returns true if person is of age or if they have permission from parents.
Create class Student with the following:
Inherits from Person.
Constructor extends parent's constructor by adding @classroom and a parameter for it.
Method play_hooky that returns "¯\(ツ)/¯".
Create class Teacher with the following:
Inherits from Person.
Constructor extends parent's constructor by adding @specialization and a parameter for it.
Override can_use_services? so it always returns true.

leonard33 committed:
Add person, student and teacher classes

leonard33 committed:
Add gemfiles

leonard33 committed:
correct linters

rustam-tolipov commented on person.rb:
```diff
@@ -0,0 +1,37 @@
+    @age = age
+  end
+
+  def gets_id
+    @id
+  end
+
+  def gets_name
+    @name
+  end
+
+  def gets_age
+    @age
+  end
+
+  def sets_name(name)
+    @name = name
+  end
+
+  def sets_age(age)
+    @age = age
+  end
```
It's better to use attribute accessors instead of creating methods for each.
Example:
```
  attr_accessor :name, :age 
  attr_reader :id 
```

>:id should only be able to be retrieved so only the Getter is defined

rustam-tolipov commented on student.rb:
```diff
@@ -0,0 +1,12 @@
+require './person'
+
+class Student < Person
+  def initialize(name, age, parent_permission, classroom)
```
I would suggest you add default values for name and parent_permission
Example:
`name = "Unknown"`
`parent_permission = true`

rustam-tolipov commented on person.rb:
```diff
@@ -0,0 +1,37 @@
+class Person
+  def initialize(name, age, _parent_permission = 'true')
```
Why did you start parent_permission with `_`?

rustam-tolipov commented on teacher.rb:
```diff
@@ -0,0 +1,12 @@
+require './person'
+
+class Teacher < Person
+  def initialize(name, age, parent_permission, specialisation)
```
Kindly give default values to name and parent_permission.

rustam-tolipov reviewed (changes requested):
Hi @leonard33, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

## Highlights
_Your project is almost complete!_
_No Linters errors_


## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

leonard33 committed:
Fix correction as per review

leonard33 committed:
fix linters

leonard33 commented:
PR- corrections
- use attribute accessors instead of creating methods for each.
- Added default values for name and parent_permission in person.rb
- Changed _ parent_permission to parent_permision
- Added default values to name and parent_permission in teachers.rb

Deepakdanger commented on person.rb:
```diff
@@ -1,29 +1,11 @@
+  attr_accessor :name, :age, :parent_permission
 
   def can_use_service?
     @age >= 18 && @parent_permission == 'true'
```
- [ ] Instead of using ` @age >= 18` , you can use directly `of_age?` private method to check.

leonard33 replied:
I dont understand what you are talking about here

Deepakdanger replied:
@leonard33 , I mean 
```
def can_use_services? 
      is_of_age?() || @parent_permission
  end
```

Deepakdanger commented on person.rb:
```diff
@@ -1,29 +1,11 @@
 class Person
-  def initialize(name, age, _parent_permission = 'true')
+  def initialize(age, name = 'Unknown', parent_permission: true)
     @id = id
     @name = name
     @age = age
   end
```
- [ ] Here, we have to define `@parent_permission = parent_permission` in initialize, Then only you can use it in the `can_use_service?` method.

leonard33 replied:
Man I really dont understand your review. what did  you mean here

Deepakdanger replied:
Here we have to define one instance variable name `parent_permission` also. For ex:-
```ruby
def initialize(age, name = "Unknown", parent_permission = true)
    @id = Random.rand(1..1000)
    @age = age
    @name = name
    @parent_permission = parent_permission
  end
```

Deepakdanger reviewed (changes requested):
Hi @leonard33, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

- N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

leonard33 committed:
Correction as per reviewer

leonard33 commented:
PR- CORRECTION
- Instead of using @age >= 18 ,  used  of_age?()  private method to check.

leonard33 committed:
Correction as per reviewer

gemmen29 reviewed (approved):
Hi @leonard33, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 💯 
- No linters errors ✔️ 
- Descriptive pull request ✔️ 
- Descriptive commit messages ✔️ 

![](https://media.giphy.com/media/mCIjCgs3nWQWfJZvPA/giphy-downsized-large.gif)

## Optional suggestions

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
